src/btree/bt_split.c

```diff
@@ -1539,45 +1566,50 @@ __split_multi_inmem_final(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *mul
 
         if (supd->ins == NULL) {
             slot = WT_ROW_SLOT(orig, supd->ripcip);
-            upd = orig->modify->mod_row_update[slot];
             orig->modify->mod_row_update[slot] = NULL;
-        } else {
-            upd = supd->ins->upd;
+        } else
             supd->ins->upd = NULL;
-        }
 
-        /*
-         * Free the onpage value and the older versions moved to the history store. We can't free
-         * the updates for in memory database and fixed length column store as they don't support
-         * the history sore.
-         */
+        /* Free the updates written to the data store and the history store. */
         if (supd->onpage_upd != NULL && !F_ISSET(S2C(session), WT_CONN_IN_MEMORY) &&
-          orig->type != WT_PAGE_COL_FIX) {
-            /*
-             * We have decided to restore this update chain so it must have newer updates than the
-             * onpage value on it.
-             */
-            WT_ASSERT(session, upd != NULL && upd != supd->onpage_upd && supd->onpage_upd != NULL);
-            /*
-             * Move the pointer to the position before the onpage value and free all the updates
-             * starting from the onpage value.
-             */
-            for (; upd->next != NULL && upd->next != supd->onpage_upd; upd = upd->next)
-                ;
-            WT_ASSERT(session, upd->next == supd->onpage_upd);
-            upd->next = NULL;
+          orig->type != WT_PAGE_COL_FIX)
             __wt_free_update_list(session, &supd->onpage_upd);
-        }
     }
 }
 
 /*
  * __split_multi_inmem_fail --
- *     Discard allocated pages after failure.
+ *     Discard allocated pages after failure and append the onpage values back to the original
+ *     update chains.
  */
 static void
-__split_multi_inmem_fail(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_REF *ref)
+__split_multi_inmem_fail(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi, WT_REF *ref)
 {
+    WT_SAVE_UPD *supd;
+    WT_UPDATE *upd;
+    uint32_t i, slot;
+
+    /* Append the onpage values back to the original update chains. */
+    for (i = 0, supd = multi->supd; i < multi->supd_entries; ++i, ++supd) {
```

Comment:
This looks correct to me, but it also seems like risky code that's probably very lightly tested. Let's keep that in mind if we see odd data consistency failures in test/format in the future (that's how I think we'd notice this going wrong).